modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java

```diff
@@ -698,22 +733,41 @@ private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitel
                 javaClass = Class.forName(whitelistStruct.javaClassName, true, whitelistClassLoader);
             } catch (ClassNotFoundException cnfe) {
                 throw new IllegalArgumentException("invalid java class name [" + whitelistStruct.javaClassName + "]" +
-                        " for struct [" + whitelistStruct.painlessTypeName + "]");
+                        " for struct [" + painlessTypeName + "]");
             }
         }
 
-        Struct existingStruct = structsMap.get(whitelistStruct.painlessTypeName);
+        Struct existingStruct = structsMap.get(painlessTypeName);
 
         if (existingStruct == null) {
-            Struct struct = new Struct(whitelistStruct.painlessTypeName, javaClass, org.objectweb.asm.Type.getType(javaClass));
+            Struct struct = new Struct(painlessTypeName, javaClass, org.objectweb.asm.Type.getType(javaClass));
+            structsMap.put(painlessTypeName, struct);
 
-            structsMap.put(whitelistStruct.painlessTypeName, struct);
-            simpleTypesMap.put(whitelistStruct.painlessTypeName, getTypeInternal(whitelistStruct.painlessTypeName));
+            if (simpleTypesMap.containsKey(painlessTypeName) == false && simpleTypesMap.containsKey(importedPainlessTypeName) == false) {
+                simpleTypesMap.put(painlessTypeName, getTypeInternal(painlessTypeName));
+            }
         } else if (existingStruct.clazz.equals(javaClass) == false) {
-            throw new IllegalArgumentException("struct [" + whitelistStruct.painlessTypeName + "] is used to " +
+            throw new IllegalArgumentException("struct [" + painlessTypeName + "] is used to " +
                     "illegally represent multiple java classes [" + whitelistStruct.javaClassName + "] and " +
                     "[" + existingStruct.clazz.getName() + "]");
         }
+
+        if (whitelistStruct.importJavaClassName) {
+            existingStruct = structsMap.get(importedPainlessTypeName);
+
+            if (existingStruct == null) {
+                structsMap.put(importedPainlessTypeName, structsMap.get(painlessTypeName));
+                Type painlessType = simpleTypesMap.remove(painlessTypeName);
+
+                if (simpleTypesMap.containsKey(importedPainlessTypeName) == false) {
+                    simpleTypesMap.put(importedPainlessTypeName, painlessType);
+                }
+            } else if (existingStruct.clazz.equals(javaClass) == false) {
+                throw new IllegalArgumentException("imported name [" + painlessTypeName + "] is used to " +
+                    "illegally represent multiple java classes [" + whitelistStruct.javaClassName + "] " +
```

Comment:
nit: I would reword this, as "illegally" sounds odd. Perhaps something like:

`Simple type name [Foo] found for types [my.Foo] and [other.Foo]. At least one must require fully qualified name using "only_fqn"`

Reply:
Changed.